packages-exp/auth-exp/src/core/auth/firebase_internal.ts

```diff
@@ -0,0 +1,86 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Unsubscribe } from '@firebase/util';
+
+import { Auth } from '../../model/auth';
+
+declare module '@firebase/component' {
+  interface NameServiceMapping {
+    'auth-internal-exp': AuthInternal;
+  }
+}
+
+interface TokenListener {
+  (tok: string | null): unknown;
+}
+
+export class AuthInternal {
+  private readonly internalListeners: Map<
+    TokenListener,
+    Unsubscribe
+  > = new Map();
+
+  constructor(private readonly auth: Auth) {}
+
+  getUid(): string | null {
+    return this.auth.currentUser?.uid || null;
+  }
+
+  async getToken(
+    forceRefresh?: boolean
+  ): Promise<{ accessToken: string } | null> {
+    await this.auth._initializationPromise;
+    if (!this.auth.currentUser) {
+      return null;
+    }
+
+    const accessToken = await this.auth.currentUser.getIdToken(forceRefresh);
+    return { accessToken };
+  }
+
+  addAuthTokenListener(listener: TokenListener): void {
+    if (this.internalListeners.has(listener)) {
+      return;
+    }
+
+    const unsubscribe = this.auth._onIdTokenChanged(user => {
+      listener(user?.stsTokenManager.accessToken || null);
+    });
+    this.internalListeners.set(listener, unsubscribe);
+    this.updateProactiveRefresh();
+  }
+
+  removeAuthTokenListener(listener: TokenListener): void {
+    const unsubscribe = this.internalListeners.get(listener);
+    if (!unsubscribe) {
+      return;
+    }
+
+    this.internalListeners.delete(listener);
+    unsubscribe();
+    this.updateProactiveRefresh();
+  }
+
+  private updateProactiveRefresh(): void {
+    if (this.internalListeners.size > 0) {
```

Comment:
Should we actively refresh for external listeners as well?

Reply:
We've definitely had feature requests for it. I think it should still be an optional thing you turn on, exposed through a method on user or something

Reply:
Are we going to do it, or this is something for the future?

Reply:
I can add it to the new API proposal and see what kind of feedback we get